python/dgl/dataloading/labor_sampler.py

```diff
@@ -0,0 +1,207 @@
+    edge type, one should instead provide a list of dicts.  Each dict would specify the
+    number of neighbors to pick per edge type.
+
+    >>> sampler = dgl.dataloading.LaborSampler([
+    ...     {('user', 'follows', 'user'): 5,
+    ...      ('user', 'plays', 'game'): 4,
+    ...      ('game', 'played-by', 'user'): 3}] * 3)
+
+    If you would like non-uniform labor sampling:
+
+    >>> g.edata['p'] = torch.rand(g.num_edges())   # any non-negative 1D vector works
+    >>> sampler = dgl.dataloading.LaborSampler([5, 10, 15], prob='p')
+
+    **Edge classification and link prediction**
+
+    This class can also work for edge classification and link prediction together
+    with :func:`as_edge_prediction_sampler`.
+
+    >>> sampler = dgl.dataloading.LaborSampler([5, 10, 15])
+    >>> sampler = dgl.dataloading.as_edge_prediction_sampler(sampler)
+    >>> dataloader = dgl.dataloading.DataLoader(
+    ...     g, train_eid, sampler,
+    ...     batch_size=1024, shuffle=True, drop_last=False, num_workers=4)
+
+    See the documentation :func:`as_edge_prediction_sampler` for more details.
+
+    Notes
+    -----
+    For the concept of MFGs, please refer to
+    :ref:`User Guide Section 6 <guide-minibatch>` and
+    :doc:`Minibatch Training Tutorials <tutorials/large/L0_neighbor_sampling_overview>`.
+    """
+
+    def __init__(
+        self,
+        fanouts,
+        edge_dir="in",
+        prob=None,
+        importance_sampling=0,
+        layer_dependency=False,
+        prefetch_node_feats=None,
+        prefetch_labels=None,
+        prefetch_edge_feats=None,
+        output_device=None,
+    ):
+        super().__init__(
+            prefetch_node_feats=prefetch_node_feats,
+            prefetch_labels=prefetch_labels,
+            prefetch_edge_feats=prefetch_edge_feats,
+            output_device=output_device,
+        )
+        self.fanouts = fanouts
+        self.edge_dir = edge_dir
+        self.prob = prob
+        self.importance_sampling = importance_sampling
+        self.layer_dependency = layer_dependency
+        self.set_seed()
+
+    def set_seed(self, random_seed=None):
+        """Updates the underlying seed for the sampler
+        
+        The passed random_seed makes it so that for any seed vertex s and its neighbor t, the rolled
+        random variate r_t is the same for any instance of this class with the same random seed.
+        When sampling as part of the same batch, one would want identical seeds so that LABOR
+        can globally sample. One example is that for heterogenous graphs, there is a single random
+        seed passed for each edge type. This will sample much fewer vertices compared to having
+        unique random seeds for each edge type. If one called this function individually for each
+        edge type for a heterogenous graph with different random seeds, then it would run LABOR
+        locally for each edge type, resulting into a larger number of vertices being sampled.
```

Comment:
The reason why LABOR samples less than neighbor sampling in general is that it uses the same random seed for each seed vertex. If each seed vertex was to use a different random seed, then labor would turn into neighbor sampling. Since the whole idea and the reduction in the number of vertices sampled comes from using identical seeds, the same thing happens in the heterogenous case as well. I don't know how to further elaborate.